packages/contracts/contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol

```diff
@@ -949,26 +1075,64 @@ contract OVM_ExecutionManager is iOVM_ExecutionManager, Lib_AddressResolver {
      * Ensures that OVM-related measures are enforced, including L2 gas refunds, nuisance gas, and flagged reversions.
      *
      * @param _nextMessageContext Message context to be used for the external message.
-     * @param _gasLimit Amount of gas to be passed into this message.
+     * @param _gasLimit Amount of gas to be passed into this message. NOTE: this argument is overwritten in some cases to avoid stack-too-deep.
      * @param _contract OVM address being called or deployed to
      * @param _data Data for the message (either calldata or creation code)
-     * @param _isCreate Whether this is a create-type message.
+     * @param _messageType What type of ovmOPCODE this message corresponds to.
      * @return Whether or not the message (either a call or deployment) succeeded.
      * @return Data returned by the message.
      */
     function _handleExternalMessage(
         MessageContext memory _nextMessageContext,
+        // NOTE: this argument is overwritten in some cases to avoid stack-too-deep.
         uint256 _gasLimit,
         address _contract,
         bytes memory _data,
-        bool _isCreate
+        MessageType _messageType
     )
         internal
         returns (
             bool,
             bytes memory
         )
     {
+        uint256 messageValue = _nextMessageContext.ovmCALLVALUE;
+        // If there is value in this message, we need to transfer the ETH over before switching contexts.
+        if (
+            messageValue > 0
+            && _isValueType(_messageType)
+        ) {
+            // Handle out-of-intrinsic gas consistent with EVM behavior -- the subcall "appears to revert" if we don't have enough gas to transfer the ETH.
+            // Similar to dynamic gas cost of value exceeding gas here:
+            // https://github.com/ethereum/go-ethereum/blob/c503f98f6d5e80e079c1d8a3601d188af2a899da/core/vm/interpreter.go#L268-L273 
+            if (gasleft() < CALL_WITH_VALUE_INTRINSIC_GAS) {
+                return (false, hex"");
+            }
+
+            // If there *is* enough gas to transfer ETH, then we need to make sure this amount of gas is reserved (i.e. not
+            // given to the _contract.call below) to guarantee that _handleExternalMessage can't run out of gas.
+            // In particular, in the event that the call fails, we will need to transfer the ETH back to the sender.
+            // Taking the lesser of _gasLimit and gasleft() - CALL_WITH_VALUE_INTRINSIC_GAS guarantees that the second
+            // _attemptForcedEthTransfer below, if needed, always has enough gas to succeed.
+            _gasLimit = Math.min(
+                _gasLimit,
+                gasleft() - CALL_WITH_VALUE_INTRINSIC_GAS // Cannot overflow due to the above check.
```

Comment:
Try cutting the subtraction in half and see if integration tests still pass

Reply:
Decided doing this is just equivalent to what we have now but harder to test, so leaving it alone.